librz/analysis/arch/arm/arm_il64.c

```diff
@@ -0,0 +1,2682 @@
+	if (!addr) {
+		return NULL;
+	}
+	addend_reg = xreg_of_reg(addend_reg);
+	RzILOpEffect *ld_eff = NULL;
+	if (OPCOUNT() == 3) {
+		// LDADD... instead of STADD, which does not have a dst reg
+		if (!ISREG(1)) {
+			rz_il_op_pure_free(addr);
+			return NULL;
+		}
+		arm64_reg dst_reg = REGID(1);
+		dst_reg = xreg_of_reg(dst_reg);
+		ld_eff = write_reg(dst_reg, loadsz != 64 ? UNSIGNED(64, VARL("old")) : VARL("old"));
+		if (!ld_eff) {
+			rz_il_op_pure_free(addr);
+			return NULL;
+		}
+	}
+	RzILOpBitVector *res = read_reg(addend_reg);
+	if (!res) {
+		rz_il_op_effect_free(ld_eff);
+		return NULL;
+	}
+	if (loadsz != 64) {
+		res = UNSIGNED(loadsz, res);
+	}
+	switch (op) {
+	case OP_CLR:
+		res = LOGAND(VARL("old"), LOGNOT(res));
+		break;
+	case OP_EOR:
+		res = LOGXOR(VARL("old"), res);
+		break;
+	case OP_SET:
+		res = LOGOR(VARL("old"), res);
+		break;
+	case OP_SMAX:
+		res = LET("r", res, ITE(SLE(VARL("old"), VARLP("r")), VARLP("r"), VARL("old")));
+		break;
+	case OP_SMIN:
+		res = LET("r", res, ITE(SLE(VARL("old"), VARLP("r")), VARL("old"), VARLP("r")));
+		break;
+	case OP_UMAX:
+		res = LET("r", res, ITE(ULE(VARL("old"), VARLP("r")), VARLP("r"), VARL("old")));
+		break;
+	case OP_UMIN:
+		res = LET("r", res, ITE(ULE(VARL("old"), VARLP("r")), VARL("old"), VARLP("r")));
+		break;
+	default: // OP_ADD
+		res = ADD(VARL("old"), res);
+		break;
+	}
+	RzILOpEffect *eff = SEQ2(
+		SETL("old", loadsz == 8 ? LOAD(addr) : LOADW(loadsz, addr)),
+		loadsz == 8 ? STORE(DUP(addr), res) : STOREW(DUP(addr), res));
+	if (ld_eff) {
+		eff = SEQ2(eff, ld_eff);
+	}
+	return eff;
+}
+#endif
+
+/**
+ * Capstone: ARM64_INS_MADD, ARM64_INS_MSUB
+ * ARM: madd, msub
+ */
+static RzILOpEffect *madd(cs_insn *insn) {
+	if (!ISREG(0)) {
+		return NULL;
+	}
+	ut32 bits = REGBITS(0);
+	RzILOpBitVector *ma = ARG(1, &bits);
+	RzILOpBitVector *mb = ARG(2, &bits);
+	RzILOpBitVector *addend = ARG(3, &bits);
+	if (!ma || !mb || !addend) {
+		return NULL;
+	}
+	RzILOpBitVector *res;
+	if (insn->id == ARM64_INS_MSUB) {
+		res = SUB(addend, MUL(ma, mb));
+	} else {
+		res = ADD(MUL(ma, mb), addend);
+	}
+	return write_reg(REGID(0), res);
+}
+
+/**
+ * Capstone: ARM64_INS_MUL, ARM64_INS_MNEG
+ * ARM: mul, mneg
+ */
+static RzILOpEffect *mul(cs_insn *insn) {
+	if (!ISREG(0)) {
+		return NULL;
+	}
+	ut32 bits = REGBITS(0);
+	RzILOpBitVector *ma = ARG(1, &bits);
+	RzILOpBitVector *mb = ARG(2, &bits);
+	if (!ma || !mb) {
+		rz_il_op_pure_free(ma);
+		rz_il_op_pure_free(mb);
+		return NULL;
+	}
+	RzILOpBitVector *res = MUL(ma, mb);
+	if (insn->id == ARM64_INS_MNEG) {
+		res = NEG(res);
+	}
+	return write_reg(REGID(0), res);
+}
+
+static RzILOpEffect *movn(cs_insn *insn);
+
+/**
+ * Capstone: ARM64_INS_MOV, ARM64_INS_MOVZ
+ * ARM: mov, movz
+ */
+static RzILOpEffect *mov(cs_insn *insn) {
+	if (!ISREG(0)) {
+		return NULL;
+	}
+	if (ISIMM(1) && IMM(1) == 0 && !strcmp(insn->mnemonic, "movn")) {
+		// Capstone bug making 0000a012 indistinguishable from 0000a052
```

Comment:
Should we open an issue to track this? And add a link here in the comment?

Reply:
Yes, maybe. I have not checked how it looks after llvm sync though.

Reply:
https://github.com/capstone-engine/capstone/issues/1857